### PR TITLE
global vs per site preference

### PR DESCRIPTION
**Pull request by jhonalino**
implementation is not yet complete. just creating a pull request for initial GUI showcase. 

image

### CODE REVIEW

**Comment by jhonalino** [comment]
image

**Comment by ansh** [comment]
@jhonalino This looks quite good! Great job so far.

File: src/style.css [hunk 1]
```diff
@@ -1,58 +1,142 @@
...
+  border: none;
+  border-radius: 4px;
+  cursor: pointer;
 
-.modeLabel {
-	display: none;
+  background: transparent;
+  color: #3aa757;
+  outline: 1px solid #3aa757;
+  box-sizing: border-box;
```
Comment (Explosion-Scratch) [inline src/style.css:19]:
```suggestion
  box-sizing: border-box;
  transition-property: background-color, transform, box-shadow;
  transition-duration: 200ms;
```

File: src/style.css [hunk 2]
```diff
@@ -1,58 +1,142 @@
...
 }
```
Comment (Explosion-Scratch) [inline src/style.css:21]:
```suggestion
button:hover, button:focus {
  box-shadow: 0 0 0 3px #3aa75733;
}
```

File: src/style.css [hunk 3]
```diff
@@ -1,58 +1,142 @@
...
-	padding: 10px;
-	min-width: 170px;
-	justify-content: center;
-	gap: 14px;
+  padding: 10px;
+  min-width: 170px;
+  justify-content: center;
+  gap: 14px;
+  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen-Sans,
+    Ubuntu, Cantarell, Helvetica Neue, sans-serif;
```
Comment (Explosion-Scratch) [inline src/style.css:7]:
```suggestion
    Ubuntu, Cantarell, Helvetica Neue, sans-serif;
  font-weight: 200;
  color: #666;
  font-size: 1.3rem;
  accent-color: #2a8241;
```

File: src/style.css [hunk 4]
```diff
@@ -1,58 +1,142 @@
...
+button:active,
+button.selected {
+  background: #3aa757;
+  outline: 1px solid transparent;
+  color: white;
 }
 
 button:active {
-	transform: scale(1.05);
+  transform: scale(1.05);
```
Comment (Explosion-Scratch) [inline src/style.css:30]: 1.05 looks a bit dramatic
```suggestion
  transform: scale(1.02);
```

**Review by Explosion-Scratch — CHANGES_REQUESTED** [review]
Maybe we could incorporate these styles as well?
- Hover + Focus shadow on buttons
- Less dramatic scale effect
- `accent-color` for inputs
- Better font, dark gray now with a light font-weight, slightly larger

**Comment by jhonalino** [comment]
Implementation + demo
https://user-images.githubusercontent.com/13143458/171586678-a79460d4-da0b-4541-a0df-bc12af4d1169.mp4
would highly appreciate feedback. thank you very much. one last thing I need to implement is the "run on pageload" but it feels like I would have to duplicate bunch of code between content and popup scripts. any idea how to go about it a cleaner way?

**Comment by asieduernest12** [comment]
i like this too looks great. will take some time to look through but i dont see anything wrong yet. will give more feedback after looking through.

**Comment by jhonalino** [comment]
implemented the page load setting
image

**Comment by jhonalino** [comment]
updated the lineheight logic so data is set from prefs -> view instead of view->prefs->view it should still function the same.

and with that update, the new ContentScript preferences logic now looks like this: 
which has access to all the preferences including the onPageLoad on subscribe 
PopupScript is using this same init API to subscribe to preference too

`  const { start } = Preferences.init({
    getOrigin: async () => new Promise((resolve, _) => {
      resolve(window.location.origin);
    }),
    subscribe: (prefs) => {
      if (!prefs.onPageLoad) {
        return;
      }
      setSaccadesIntervalInDOM(prefs.saccadesInterval);
      setReadingMode(prefs.enabled);
      setFixationStrength(prefs.fixationStrength);
      setLineHeight(prefs.lineHeight);
    },
  });
` 

and with this update. I believe this pull request implementation is complete. Looking forward for your feedback. thank you very much!

**Review by asieduernest12 — COMMENTED** [review]
looks really impressive some comments here and there expecially summarizing the logic would go a long way to help anyone understand whats going on here without looking at the code so hard my screen feels offended. but i really like this. nice work... high five

**Comment by jhonalino** [comment]
> looks really impressive. some comments here and there mosly summarizing the logic would go a long way to help anyone understand whats going on here (preferences/index.js) without looking at the code so hard my screen feels offended.
> 
> but i really like this. nice work... high five

high five!
thank you so much for the feedback! I've added comments all the functions for preferences/index.js. please do let me know if those comments do provide clarity or if its still missing some explanations to help anyone understand what's going on 

>  without looking at the code so hard my screen feels offended.
> 
hahaha my bad !

**Comment by asieduernest12** [comment]
other issues
[on pageload not acting as expected](https://youtu.be/eytJIQJTBkI)

[default fixations](https://youtu.be/sKHQAH7iM3k)

[site preferences proposal](https://youtu.be/liJCXGBMmpA)

**Review by asieduernest12 — CHANGES_REQUESTED** [review]

**Comment by Explosion-Scratch** [comment]
> 

You know you can just drag + drop the MP4 files onto an issue comment right?

**Comment by jhonalino** [comment]
> other issues in video form
> 
> [on pageload not acting as expected](https://youtu.be/eytJIQJTBkI)
> 
> [default fixations](https://youtu.be/sKHQAH7iM3k)
> 
> [site preferences proposal](https://youtu.be/liJCXGBMmpA)

@asieduernest12 you're the best! I highly appreciate these videos feedback. I've committed updates to address the first 2 issue.  
1 [on pageload not acting as expected](https://youtu.be/eytJIQJTBkI)

2 [default fixations](https://youtu.be/sKHQAH7iM3k)

and will work on the third one next.

**Comment by asieduernest12** [comment]
awesome awesome 
so i think there is one more buggy behaviour with the reading mode toggle
[toggle buggy behaviour video](https://youtu.be/DdvnQmN40vE)

it is also present when you use the site preferences

**Comment by Explosion-Scratch** [comment]
> [toggle buggy behaviour video](https://youtu.be/DdvnQmN40vE)

That sounds like a mix of an autogenerated title and a tongue twister

**Comment by asieduernest12** [comment]
> > 
> 
> You know you can just drag + drop the MP4 files onto an issue comment right?

the videos must be under 10mb to be added to a comment, the earlier videos were above the limit

**Comment by Explosion-Scratch** [comment]
> > > 
> > 
> > 
> > You know you can just drag + drop the MP4 files onto an issue comment right?
> 
> the videos must be under 10mb to be added to a comment, the earlier videos were above the limit

My screen recorder produces massive files even though my screen resolution is not incredible and simply converting them to MP4 and increasing the speed 2x really compresses them (like 80%)

**Comment by asieduernest12** [comment]
> > [toggle buggy behaviour video](https://youtu.be/DdvnQmN40vE)
> 
> That sounds like a mix of an autogenerated title and a tongue twister
HappyBabyLaughingGIF

**Comment by asieduernest12** [comment]
> > > > 
> > > 
> > > 
> > > You know you can just drag + drop the MP4 files onto an issue comment right?
> > 
> > 
> > the videos must be under 10mb to be added to a comment, the earlier videos were above the limit
> 
> My screen recorder produces massive files even though my screen resolution is not incredible and simply converting them to MP4 and increasing the speed 2x really compresses them (like 80%)



> > > > 
> > > 
> > > 
> > > You know you can just drag + drop the MP4 files onto an issue comment right?
> > 
> > 
> > the videos must be under 10mb to be added to a comment, the earlier videos were above the limit
> 
> My screen recorder produces massive files even though my screen resolution is not incredible and simply converting them to MP4 and increasing the speed 2x really compresses them (like 80%)

yeah the speed up trick works magic but i cant do that since i was talking in the video. speeding it up would just turn me into a horrible tongue twister rambling

**Comment by jhonalino** [comment]
> awesome awesome so i think there is one more buggy behaviour with the reading mode toggle [toggle buggy behaviour video](https://youtu.be/DdvnQmN40vE)
> 
> it is also present when you use the site preferences

thank you so much for the video! I've published an update to address this issue. 
I removed that reading mode state from preferences but keep its initiate state sync only if page load is true. so if its starts out toggled via page load. it will be highlighted but after that. it can be toggled on and off without affecting page load button highlight and vice versa.  

please lmk if still causing an issue for you? thank you very much! I will continue on the [site preferences proposal](https://youtu.be/liJCXGBMmpA) next

**Comment by asieduernest12** [comment]
i think everything checks out.. thank you so much @jhonalino your code is real elegant.

**Comment by jhonalino** [comment]
@asieduernest12 I believe I got [site preferences proposal](https://youtu.be/liJCXGBMmpA) working. do you mind testing it to see if I got it right or if I'm missing something? thank you very much! highly appreciate your feedback @asieduernest12

**Review by jhonalino — COMMENTED** [review]
added parsenode back outside the bolded element check @asieduernest12 sorry about that. I thought I needed to protect that from double parsing, turns out we already have ignore logic for double parsing! thank you very much!

**Comment by Explosion-Scratch** [comment]
Cool gif

**Pull request merged**